File: server/typescript/packages/render/src/recover/strip.ts
// Stage 1: remove markdown code-fence markers. Prose around the payload is left for Locate.
// Mirrors Java Strip.

// Captures the body inside a fenced block; optional language tag (json/xml/etc) is dropped.
const FENCE = /```[a-zA-Z0-9_-]*\s*\r?\n([\s\S]*?)\r?\n?```/;

export function strip(raw: string | null | undefined): string {
  if (raw == null) return "";
  const m = FENCE.exec(raw);
  if (m && m.index >= 0) {
    const before = raw.substring(0, m.index);
    const body = m[1] ?? "";
    const after = raw.substring(m.index + m[0].length);
    return (before + body + after).trim();
  }
  return raw.trim();
}

File: server/java/render/src/main/java/com/metaobjects/render/extract/Strip.java
package com.metaobjects.render.extract;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Stage 1: remove markdown code-fence markers. Prose around the payload is left for Locate. */
public final class Strip {
    private Strip() {}

    // Captures the body inside a fenced block; optional language tag (json/xml/etc) is dropped.
    private static final Pattern FENCE = Pattern.compile(
            "```[a-zA-Z0-9_-]*\\s*\\r?\\n(.*?)\\r?\\n?```",
            Pattern.DOTALL);

    /** The body of every fenced block, in order. Locate searches these before the whole text,
     *  because a model told to fence its answer puts the answer there. */
    public static List<String> fencedBodies(String raw) {
        List<String> out = new ArrayList<>();
        if (raw == null) return out;
        Matcher m = FENCE.matcher(raw);
        while (m.find()) out.add(m.group(1));
        return out;
    }

    public static String strip(String raw) {
        if (raw == null) return "";
        Matcher m = FENCE.matcher(raw);
        if (m.find()) {
            return (raw.substring(0, m.start()) + m.group(1) + raw.substring(m.end())).trim();
        }
        return raw.trim();
    }
}
